File: tools/xrai-decoder.js
/**
 * XRAI Decoder
 * Parses XRAI format files and extracts their components
 */

class XRAIDecoder {
  constructor(options = {}) {
    this.options = {
      useCache: true,
      ...options
    };
    
    this.cache = new Map();
  }
  
  /**
   * Decode an XRAI file
   * @param {ArrayBuffer} buffer - The binary XRAI data
   * @returns {Object} - The decoded XRAI content
   */
  async decode(buffer) {
    try {
      // Check cache if enabled
      if (this.options.useCache) {
        const cacheKey = this._getCacheKey(buffer);
        if (this.cache.has(cacheKey)) {
          return this.cache.get(cacheKey);
        }
      }
      
      // Parse header
      const header = this._parseHeader(buffer);
      
      // Validate magic number
      if (header.magic !== 'XRAI') {
        throw new Error('Invalid XRAI file: incorrect magic number');
      }
      
      // Extract sections
      const sections = this._extractSections(buffer, header.tocOffset);
      
      // Construct result object
      const result = {
        version: header.version,
        flags: header.flags,
        ...sections
      };
      
      // Cache result if enabled
      if (this.options.useCache) {
        const cacheKey = this._getCacheKey(buffer);
        this.cache.set(cacheKey, result);
      }
      
      return result;
    } catch (error) {
      throw new Error(`XRAI decoding failed: ${error.message}`);
    }
  }
  
  /**
   * Validate an XRAI file
   * @param {ArrayBuffer} buffer - The binary XRAI data
   * @returns {Object} - Validation result
   */
  validate(buffer) {
    try {
      // Parse header
      const header = this._parseHeader(buffer);
      
      // Check magic number
      if (header.magic !== 'XRAI') {
        return {
          valid: false,
          errors: ['Invalid magic number']
        };
      }
      
      // Check version compatibility
      if (header.version.major > 1) {
        return {
          valid: false,
          errors: [`Unsupported version: ${header.version.major}.${header.version.minor}`]
        };
      }
      
      // Check TOC
      try {
        const sections = this._extractSections(buffer, header.tocOffset, true);
        
        // Check required sections
        const requiredSections = ['metadata'];
        const missingSections = requiredSections.filter(section => !sections[section]);
        
        if (missingSections.length > 0) {
          return {
            valid: false,
            errors: [`Missing required sections: ${missingSections.join(', ')}`]
          };
        }
        
        return {
          valid: true,
          version: header.version,
          sections: Object.keys(sections)
        };
      } catch (error) {
        return {
          valid: false,
          errors: [`Invalid table of contents: ${error.message}`]
        };
      }
    } catch (error) {
      return {
        valid: false,
        errors: [`Validation failed: ${error.message}`]
      };
    }
  }
  
  /**
   * Parse the XRAI header
   * @private
   * @param {ArrayBuffer} buffer - The binary XRAI data
   * @returns {Object} - The parsed header
   */
  _parseHeader(buffer) {
    const headerView = new DataView(buffer);
    
    // Read magic number (4 bytes)
    const magic = String.fromCharCode(
      headerView.getUint8(0),
      headerView.getUint8(1),
      headerView.getUint8(2),
      headerView.getUint8(3)
    );
    
    // Read version (2 bytes)
    const version = {
      major: headerView.getUint8(4),
      minor: headerView.getUint8(5)
    };
    
    // Read flags (2 bytes)
    const flags = headerView.getUint16(6, true);
    
    // Read TOC offset (8 bytes)
    const tocOffset = Number(headerView.getBigUint64(8, true));
    
    return { magic, version, flags, tocOffset };
  }
  
  /**
   * Extract sections from the XRAI file
   * @private
   * @param {ArrayBuffer} buffer - The binary XRAI data
   * @param {number} tocOffset - Offset to the table of contents
   * @param {boolean} validateOnly - If true, only validate without parsing section data
   * @returns {Object} - The extracted sections
   */
  _extractSections(buffer, tocOffset, validateOnly = false) {
    const tocView = new DataView(buffer, tocOffset);
    const sectionCount = tocView.getUint32(0, true);
    
    // Check if section count is reasonable
    if (sectionCount > 100) {
      throw new Error(`Invalid section count: ${sectionCount}`);
    }
    
    const sections = {};
    let offset = 4;
    
    for (let i = 0; i < sectionCount; i++) {
      const typeId = tocView.getUint32(offset, true);
      const sectionOffset = Number(tocView.getBigUint64(offset + 4, true));
      const sectionSize = Number(tocView.getBigUint64(offset + 12, true));
      const sectionFlags = tocView.getUint32(offset + 20, true);
      
      // Check if section offset and size are valid
      if (sectionOffset + sectionSize > buffer.byteLength) {
        throw new Error(`Invalid section bounds: offset=${sectionOffset}, size=${sectionSize}`);
      }
      
      const sectionType = this._getSectionTypeFromId(typeId);
      
      if (!validateOnly) {
        const sectionData = buffer.slice(sectionOffset, sectionOffset + sectionSize);
        sections[sectionType] = this._parseSection(sectionType, sectionData, sectionFlags);
      } else {
        sections[sectionType] = true; // Just mark as present for validation
      }
      
      offset += 24;
    }
    
    return sections;
  }
  
  /**
   * Parse a section based on its type
   * @private
   * @param {string} type - The section type
   * @param {ArrayBuffer} data - The section data
   * @param {number} flags - The section flags
   * @returns {Object|Array} - The parsed section data
   */
  _parseSection(type, data, flags) {
    // Check if data is compressed
    const isCompressed = (flags & 0x1) !== 0;
    
    // Decompress if needed
    const decompressedData = isCompressed ? this._decompressData(data, flags) : data;
    
    // Parse based on section type
    switch (type) {
      case 'metadata':
        return this._parseMetadata(decompressedData);
      case 'geometry':
        return this._parseGeometry(decompressedData);
      case 'materials':
        return this._parseMaterials(decompressedData);
      case 'animations':
        return this._parseAnimations(decompressedData);
      case 'audio':
        return this._parseAudio(decompressedData);
      case 'aiComponents':
        return this._parseAIComponents(decompressedData);
      case 'vfx':
        return this._parseVFX(decompressedData);
      default:
        console.warn(`Unknown section type: ${type}`);
        return decompressedData;
    }
  }
  
  /**
   * Get section type from type ID
   * @private
   * @param {number} typeId - The section type ID
   * @returns {string} - The section type name
   */
  _getSectionTypeFromId(typeId) {
    const types = {
      1: 'metadata',
      2: 'geometry',
      3: 'materials',
      4: 'animations',
      5: 'audio',
      6: 'aiComponents',
      7: 'vfx',
      // Add more section types as needed
    };
    
    return types[typeId] || `unknown_${typeId}`;
  }
  
  /**
   * Decompress data
   * @private
   * @param {ArrayBuffer} data - The compressed data
   * @param {number} flags - The section flags
   * @returns {ArrayBuffer} - The decompressed data
   */
  _decompressData(data, flags) {
    // Extract compression algorithm from flags
    const compressionAlgo = (flags >> 8) & 0xFF;
    
    // Implement decompression based on algorithm
    // This is a placeholder - actual implementation would depend on the compression used
    console.log(`Decompressing data with algorithm ${compressionAlgo}`);
    
    // For this example, just return the original data
    return data;
  }
  
  /**
   * Parse metadata section
   * @private
   * @param {ArrayBuffer} data - The metadata section data
   * @returns {Object} - The parsed metadata
   */
  _parseMetadata(data) {
    // Convert ArrayBuffer to string
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    // Parse JSON
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing metadata JSON:', e);
      return {};
    }
  }
  
  /**
   * Parse geometry section
   * @private
   * @param {ArrayBuffer} data - The geometry section data
   * @returns {Array} - The parsed geometry data
   */
  _parseGeometry(data) {
    // This is a simplified placeholder
    // Actual implementation would parse the binary geometry data
    
    // For this example, assume the geometry is stored as JSON
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing geometry data:', e);
      return [];
    }
  }
  
  /**
   * Parse materials section
   * @private
   * @param {ArrayBuffer} data - The materials section data
   * @returns {Array} - The parsed materials data
   */
  _parseMaterials(data) {
    // Similar to geometry parsing, this is a simplified placeholder
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing materials data:', e);
      return [];
    }
  }
  
  /**
   * Parse animations section
   * @private
   * @param {ArrayBuffer} data - The animations section data
   * @returns {Array} - The parsed animations data
   */
  _parseAnimations(data) {
    // Simplified placeholder
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing animations data:', e);
      return [];
    }
  }
  
  /**
   * Parse audio section
   * @private
   * @param {ArrayBuffer} data - The audio section data
   * @returns {Object} - The parsed audio data
   */
  _parseAudio(data) {
    // Simplified placeholder
    // In a real implementation, this would parse audio data and metadata
    return {
      clips: [],
      spatial: {}
    };
  }
  
  /**
   * Parse AI components section
   * @private
   * @param {ArrayBuffer} data - The AI components section data
   * @returns {Object} - The parsed AI components data
   */
  _parseAIComponents(data) {
    // Simplified placeholder
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing AI components data:', e);
      return {
        adaptationRules: [],
        behaviorModels: []
      };
    }
  }
  
  /**
   * Parse VFX section
   * @private
   * @param {ArrayBuffer} data - The VFX section data
   * @returns {Object} - The parsed VFX data
   */
  _parseVFX(data) {
    // Simplified placeholder
    const decoder = new TextDecoder('utf-8');
    const jsonString = decoder.decode(data);
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.error('Error parsing VFX data:', e);
      return {
        particleSystems: [],
        vfxGraphs: []
      };
    }
  }
  
  /**
   * Generate a cache key for a buffer
   * @private
   * @param {ArrayBuffer} buffer - The buffer to generate a key for
   * @returns {string} - The cache key
   */
  _getCacheKey(buffer) {
    // Simple hash function for ArrayBuffer
    // In a real implementation, use a proper hash function
    let hash = 0;
    const view = new Uint8Array(buffer);
    
    // Sample the buffer at regular intervals
    const step = Math.max(1, Math.floor(view.length / 100));
    for (let i = 0; i < view.length; i += step) {
      hash = ((hash << 5) - hash) + view[i];
      hash |= 0; // Convert to 32-bit integer
    }
    
    return `${hash}_${buffer.byteLength}`;
  }
  
  /**
   * Clear the decoder cache
   */
  clearCache() {
    this.cache.clear();
  }
}

module.exports = { XRAIDecoder };

File: demos/unity/XRAIDecoder.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.IO.Compression;
using UnityEngine;
using Newtonsoft.Json;

namespace XRAI
{
    /// <summary>
    /// Unity implementation of the XRAI format decoder
    /// </summary>
    public class XRAIDecoder
    {
        // Section type mapping for faster lookups
        private static readonly Dictionary<int, string> SectionTypes = new Dictionary<int, string>
        {
            { 1, "metadata" },
            { 2, "geometry" },
            { 3, "materials" },
            { 4, "animations" },
            { 5, "audio" },
            { 6, "aiComponents" },
            { 7, "vfx" },
            { 8, "buffers" },
            { 9, "images" },
            { 10, "scene" },
            { 11, "extensions" }
        };

        // Decoder options
        private bool validateOnLoad = true;
        private bool useCache = true;

        // Cache for decoded data
        private Dictionary<string, object> cache = new Dictionary<string, object>();

        /// <summary>
        /// Decode an XRAI file from a file path
        /// </summary>
        /// <param name="filePath">Path to XRAI file</param>
        /// <returns>Decoded data as a Dictionary</returns>
        public Dictionary<string, object> Decode(string filePath)
        {
            Debug.Log($"Decoding XRAI file: {filePath}");
            
            // Check if file exists
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException($"XRAI file not found: {filePath}");
            }
            
            // Read file as binary
            byte[] fileData = File.ReadAllBytes(filePath);
            
            return DecodeBytes(fileData);
        }
        
        /// <summary>
        /// Decode XRAI data from a byte array
        /// </summary>
        /// <param name="data">XRAI data as byte array</param>
        /// <returns>Decoded data as a Dictionary</returns>
        public Dictionary<string, object> DecodeBytes(byte[] data)
        {
            // Check cache if enabled
            if (useCache)
            {
                string cacheKey = GetCacheKey(data);
                if (cache.ContainsKey(cacheKey))
                {
                    return cache[cacheKey] as Dictionary<string, object>;
                }
            }
            
            // Validate if enabled
            if (validateOnLoad)
            {
                ValidationResult validationResult = ValidateBytes(data);
                if (!validationResult.Valid)
                {
                    throw new Exception($"Invalid XRAI file: {string.Join(", ", validationResult.Errors)}");
                }
            }
            
            // Parse header
            XRAIHeader header = ParseHeader(data);
            
            // Parse TOC
            XRAITOC toc = ParseTOC(data, header.TOCOffset);
            
            // Parse sections
            Dictionary<string, object> result = ParseSections(data, toc, header);
            
            // Add format info
            Dictionary<string, object> formatInfo = new Dictionary<string, object>
            {
                { "version", $"{header.VersionMajor}.{header.VersionMinor}" },
                { "flags", header.Flags }
            };
            result["_format"] = formatInfo;
            
            // Cache result if enabled
            if (useCache)
            {
                string cacheKey = GetCacheKey(data);
                cache[cacheKey] = result;
            }
            
            return result;
        }
        
        /// <summary>
        /// Validate XRAI data
        /// </summary>
        /// <param name="data">XRAI data as byte array</param>
        /// <returns>Validation result</returns>
        public ValidationResult ValidateBytes(byte[] data)
        {
            try
            {
                if (data.Length < 16)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { "File too small to be a valid XRAI file" }
                    };
                }
                
                // Read magic number
                string magic = Encoding.ASCII.GetString(data, 0, 4);
                
                // Validate magic number
                if (magic != "XRAI")
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { $"Invalid magic number: {magic}" }
                    };
                }
                
                // Read version
                byte versionMajor = data[4];
                byte versionMinor = data[5];
                
                // Check version
                if (versionMajor > 1)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { $"Unsupported version: {versionMajor}.{versionMinor}" }
                    };
                }
                
                // Read TOC offset
                long tocOffset = BitConverter.ToInt64(data, 8);
                
                // Check TOC offset bounds
                if (tocOffset >= data.Length)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { $"Invalid TOC offset: {tocOffset}, fileSize={data.Length}" }
                    };
                }
                
                // Read TOC
                if (tocOffset + 4 > data.Length)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { "Invalid TOC: not enough bytes to read section count" }
                    };
                }
                
                int sectionCount = BitConverter.ToInt32(data, (int)tocOffset);
                
                // Check if section count is reasonable
                if (sectionCount > 100)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { $"Invalid section count: {sectionCount} (max 100)" }
                    };
                }
                
                // Check if we can read all TOC entries
                long tocSize = 4 + (sectionCount * 24);
                if (tocOffset + tocSize > data.Length)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { "Invalid TOC: not enough bytes to read all section entries" }
                    };
                }
                
                // Validate each section
                List<SectionInfo> sections = new List<SectionInfo>();
                long tocEntryOffset = tocOffset + 4; // Skip section count
                
                for (int i = 0; i < sectionCount; i++)
                {
                    int typeId = BitConverter.ToInt32(data, (int)tocEntryOffset);
                    long sectionOffset = BitConverter.ToInt64(data, (int)tocEntryOffset + 4);
                    long sectionSize = BitConverter.ToInt64(data, (int)tocEntryOffset + 12);
                    
                    // Check section bounds
                    if (sectionOffset + sectionSize > data.Length)
                    {
                        return new ValidationResult
                        {
                            Valid = false,
                            Errors = new List<string> { $"Invalid section bounds: section {i}, offset={sectionOffset}, size={sectionSize}, fileSize={data.Length}" }
                        };
                    }
                    
                    sections.Add(new SectionInfo
                    {
                        TypeId = typeId,
                        Offset = sectionOffset,
                        Size = sectionSize
                    });
                    
                    tocEntryOffset += 24; // Size of each TOC entry
                }
                
                // Check for required sections
                bool hasMetadata = sections.Exists(section => section.TypeId == 1);
                if (!hasMetadata)
                {
                    return new ValidationResult
                    {
                        Valid = false,
                        Errors = new List<string> { "Missing required metadata section (type 1)" }
                    };
                }
                
                return new ValidationResult
                {
                    Valid = true,
                    Version = $"{versionMajor}.{versionMinor}",
                    Sections = sections
                };
            }
            catch (Exception ex)
            {
                return new ValidationResult
                {
                    Valid = false,
                    Errors = new List<string> { $"Validation failed: {ex.Message}" }
                };
            }
        }
        
        /// <summary>
        /// Parse XRAI header
        /// </summary>
        /// <param name="data">XRAI data</param>
        /// <returns>Parsed header</returns>
        private XRAIHeader ParseHeader(byte[] data)
        {
            // Read magic number
            string magic = Encoding.ASCII.GetString(data, 0, 4);
            
            // Read version
            byte versionMajor = data[4];
            byte versionMinor = data[5];
            
            // Read flags
            ushort flags = BitConverter.ToUInt16(data, 6);
            
            // Read TOC offset
            long tocOffset = BitConverter.ToInt64(data, 8);
            
            return new XRAIHeader
            {
                Magic = magic,
                VersionMajor = versionMajor,
                VersionMinor = versionMinor,
                Flags = flags,
                TOCOffset = tocOffset
            };
        }
        
        /// <summary>
        /// Parse XRAI TOC
        /// </summary>
        /// <param name="data">XRAI data</param>
        /// <param name="tocOffset">TOC offset</param>
        /// <returns>Parsed TOC</returns>
        private XRAITOC ParseTOC(byte[] data, long tocOffset)
        {
            // Read section count
            int sectionCount = BitConverter.ToInt32(data, (int)tocOffset);
            
            // Parse sections
            List<SectionEntry> sections = new List<SectionEntry>();
            long tocEntryOffset = tocOffset + 4; // Skip section count
            
            for (int i = 0; i < sectionCount; i++)
            {
                int typeId = BitConverter.ToInt32(data, (int)tocEntryOffset);
                long sectionOffset = BitConverter.ToInt64(data, (int)tocEntryOffset + 4);
                long sectionSize = BitConverter.ToInt64(data, (int)tocEntryOffset + 12);
                uint sectionFlags = BitConverter.ToUInt32(data, (int)tocEntryOffset + 20);
                
                sections.Add(new SectionEntry
                {
                    TypeId = typeId,
                    Offset = sectionOffset,
                    Size = sectionSize,
                    Flags = sectionFlags
                });
                
                tocEntryOffset += 24; // Size of each TOC entry
            }
            
            return new XRAITOC
            {
                SectionCount = sectionCount,
                Sections = sections,
                Size = 4 + (sectionCount * 24) // TOC size
            };
        }
        
        /// <summary>
        /// Parse XRAI sections
        /// </summary>
        /// <param name="data">XRAI data</param>
        /// <param name="toc">Parsed TOC</param>
        /// <param name="header">Parsed header</param>
        /// <returns>Parsed sections</returns>
        private Dictionary<string, object> ParseSections(byte[] data, XRAITOC toc, XRAIHeader header)
        {
            Dictionary<string, object> result = new Dictionary<string, object>();
            
            // Parse each section
            foreach (SectionEntry section in toc.Sections)
            {
                // Get section data
                byte[] sectionData = new byte[section.Size];
                Array.Copy(data, section.Offset, sectionData, 0, section.Size);
                
                // Get section type
                string sectionType = SectionTypes.ContainsKey(section.TypeId) 
                    ? SectionTypes[section.TypeId] 
                    : $"unknown_{section.TypeId}";
                
                // Parse section
                object sectionObj = ParseSection(sectionType, sectionData, section.Flags);
                
                // Add to result
                if (section.TypeId == 1) // Metadata
                {
                    // Merge metadata with result
                    if (sectionObj is Dictionary<string, object> metadataDict)
                    {
                        foreach (var kvp in metadataDict)
                        {
                            result[kvp.Key] = kvp.Value;
                        }
                    }
                }
                else
                {
                    result[sectionType] = sectionObj;
                }
            }
            
            return result;
        }
        
        /// <summary>
        /// Parse a section
        /// </summary>
        /// <param name="type">Section type</param>
        /// <param name="data">Section data</param>
        /// <param name="flags">Section flags</param>
        /// <returns>Parsed section data</returns>
        private object ParseSection(string type, byte[] data, uint flags)
        {
            // Check if data is compressed
            bool isCompressed = (flags & 0x1) != 0;
            int compressionAlgo = (int)((flags >> 8) & 0xFF);
            
            // Decompress if needed
            byte[] decompressedData = data;
            if (isCompressed)
            {
                try
                {
                    Debug.Log($"Decompressing {type} section: {data.Length} bytes, flags: {flags:X}, algo: {compressionAlgo}");
                    
                    // Default to deflate if no algorithm is specified or if it's explicitly set to 1
                    if (compressionAlgo == 0 || compressionAlgo == 1) // Deflate
                    {
                        using (MemoryStream compressedStream = new MemoryStream(data))
                        using (DeflateStream deflateStream = new DeflateStream(compressedStream, CompressionMode.Decompress))
                        using (MemoryStream resultStream = new MemoryStream())
                        {
                            deflateStream.CopyTo(resultStream);
                            decompressedData = resultStream.ToArray();
                        }
                        
                        Debug.Log($"Decompressed to {decompressedData.Length} bytes");
                    }
                    else
                    {
                        Debug.LogWarning($"Unknown compression algorithm: {compressionAlgo}, treating as uncompressed");
                    }
                }
                catch (Exception ex)
                {
                    Debug.LogError($"Decompression failed for {type} section: {ex.Message}");
                    // Continue with compressed data
                }
            }
            
            // Parse based on section type
            switch (type)
            {
                case "metadata":
                case "geometry":
                case "materials":
                case "animations":
                case "aiComponents":
                case "vfx":
                case "scene":
                    // Parse as JSON
                    return ParseJSON(decompressedData);
                
                case "buffers":
                case "images":
                case "audio":
                    // Return as binary data
                    return decompressedData;
                
                default:
                    // Try to parse as JSON, fall back to binary
                    try
                    {
                        return ParseJSON(decompressedData);
                    }
                    catch
                    {
                        return decompressedData;
                    }
            }
        }
        
        /// <summary>
        /// Parse JSON data
        /// </summary>
        /// <param name="data">JSON data as byte array</param>
        /// <returns>Parsed JSON</returns>
        private object ParseJSON(byte[] data)
        {
            string jsonString = Encoding.UTF8.GetString(data);
            
            try
            {
                return JsonConvert.DeserializeObject<Dictionary<string, object>>(jsonString);
            }
            catch (Exception ex)
            {
                throw new Exception($"Invalid JSON data: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Get cache key for data
        /// </summary>
        /// <param name="data">Data to get cache key for</param>
        /// <returns>Cache key</returns>
        private string GetCacheKey(byte[] data)
        {
            // Use data length and first few bytes as cache key
            StringBuilder keyBuilder = new StringBuilder();
            keyBuilder.Append(data.Length);
            keyBuilder.Append(':');
            
            for (int i = 0; i < Math.Min(16, data.Length); i++)
            {
                keyBuilder.Append(data[i].ToString("X2"));
            }
            
            return keyBuilder.ToString();
        }
        
        /// <summary>
        /// Clear decoder cache
        /// </summary>
        public void ClearCache()
        {
            cache.Clear();
        }
    }

    /// <summary>
    /// XRAI header structure
    /// </summary>
    public class XRAIHeader
    {
        public string Magic { get; set; }
        public byte VersionMajor { get; set; }
        public byte VersionMinor { get; set; }
        public ushort Flags { get; set; }
        public long TOCOffset { get; set; }
    }

    /// <summary>
    /// XRAI TOC structure
    /// </summary>
    public class XRAITOC
    {
        public int SectionCount { get; set; }
        public List<SectionEntry> Sections { get; set; }
        public long Size { get; set; }
    }

    /// <summary>
    /// XRAI section entry
    /// </summary>
    public class SectionEntry
    {
        public int TypeId { get; set; }
        public long Offset { get; set; }
        public long Size { get; set; }
        public uint Flags { get; set; }
    }

    /// <summary>
    /// XRAI section info for validation
    /// </summary>
    public class SectionInfo
    {
        public int TypeId { get; set; }
        public long Offset { get; set; }
        public long Size { get; set; }
    }

    /// <summary>
    /// XRAI validation result
    /// </summary>
    public class ValidationResult
    {
        public bool Valid { get; set; }
        public string Version { get; set; }
        public List<string> Errors { get; set; } = new List<string>();
        public List<SectionInfo> Sections { get; set; } = new List<SectionInfo>();
    }
}
